### PR TITLE
feat(console): preserve state inside resource interaction views

### PR DESCRIPTION
**Pull request by polamoros**
Preserve inputs state for the interaction views when:
* Changing from one resource to another
* Saving the file (if the resource still exists)

https://github.com/winglang/wing/assets/5547636/b5161eae-f180-4d96-bd61-63f5f051765a

resolves https://github.com/winglang/wing/issues/3928

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: apps/wing-console/console/ui/src/layout/default-layout.tsx [hunk 1]
```diff
@@ -239,6 +245,40 @@ export const DefaultLayout = ({
...
+    }
+
+    const newInstances = { ...metadataInstances };
+    for (const nodeId of Object.keys(metadataInstances)) {
+      if (!hasChild(node, nodeId)) {
+        delete newInstances[nodeId];
+      }
+    }
+    setMetadataInstances(newInstances);
+  }, [items]);
```
Comment (skyrpex) [inline apps/wing-console/console/ui/src/layout/default-layout.tsx:280]: You should pass a callback inside `setMetadataInstances` or otherwise it won't correctly work since the effect depends on `metadataInstances`

Comment (polamoros) [reply apps/wing-console/console/ui/src/layout/default-layout.tsx]: you're right!

File: apps/wing-console/console/server/src/router/app.ts [hunk 2]
```diff
@@ -114,6 +114,34 @@ export const createAppRouter = () => {
           input?.showTests,
         );
       }),
+    "app.itemIds": createProcedure
```
Comment (skyrpex) [inline apps/wing-console/console/server/src/router/app.ts:117]: Can we name it `app.nodeIds`? They are only called items in the context of a tree view, but here they are nodes.

File: apps/wing-console/console/server/src/router/app.ts [hunk 3]
```diff
@@ -114,6 +114,34 @@ export const createAppRouter = () => {
...
+        const getItemIds = (item: ExplorerItem, list: string[] = []) => {
+          list.push(item.id);
+          if (item.childItems) {
+            for (const child of item.childItems) {
+              getItemIds(child, list);
+            }
+          }
+          return list;
+        };
+        return getItemIds(node);
```
Comment (skyrpex) [inline apps/wing-console/console/server/src/router/app.ts:143]:
```suggestion
        const list = new Array<string>();
        const getItemIds = (item: ExplorerItem) => {
          list.push(item.id);
          for (const child of item.childItems ?? []) {
            getItemIds(child);
          }
        };
        getItemIds(node);
        return list;
```

I think this is more readable.

File: apps/wing-console/console/ui/src/layout/use-layout.tsx [hunk 4]
```diff
@@ -105,6 +106,7 @@ export const useLayout = ({
     },
     {
       enabled: selectedItems.length > 0,
+      keepPreviousData: false,
```
Comment (skyrpex) [inline apps/wing-console/console/ui/src/layout/use-layout.tsx:109]: Why this change?

Comment (polamoros) [reply apps/wing-console/console/ui/src/layout/use-layout.tsx:109]: I added this to prevent the UI from displaying the metadata of the previously selected node when switching from one resource to another. `useLayout` is providing the trpc object directly, and we are using the `metadata.data` in the UI, so we need to clean it up.

We could do other things like check if it's loading I guess..

with:
https://github.com/winglang/wing/assets/5547636/f2d5334e-8e66-4574-8029-352e3ec356db

without:
https://github.com/winglang/wing/assets/5547636/1b69808a-116b-4e75-87f8-08cf3205dc6a

File: apps/wing-console/console/server/src/router/app.ts (hunk 2 above)
Comment (polamoros) [reply apps/wing-console/console/server/src/router/app.ts]: yep

File: apps/wing-console/console/ui/src/layout/use-layout.tsx (hunk 4 above)
Comment (skyrpex) [reply apps/wing-console/console/ui/src/layout/use-layout.tsx:109]: Makes sense what you're doing, although I think there's something strange. Even with preserve enabled, there's a flash where the panel is empty 🤔 

Also, when did it get so slow?

Comment (polamoros) [reply apps/wing-console/console/ui/src/layout/use-layout.tsx]: It's 'slow' as usual, but now the metadata pane was hidden while fetching new data. 
I fixed this by keeping the previos metadata panel until the new metadata is fetched.

File: apps/wing-console/console/ui/src/layout/default-layout.tsx [hunk 5]
```diff
@@ -239,6 +245,43 @@ export const DefaultLayout = ({
     ],
   );
 
+  const [currentNodeMetadata, setCurrentNodeMetadata] = useState<
```
Comment (ainvoner) [inline apps/wing-console/console/ui/src/layout/default-layout.tsx:248]: isn't it just currentNodeId? ```const currentNode = selectedItems[0]``` 
 it is not the current node metadata

File: apps/wing-console/console/ui/src/layout/default-layout.tsx [hunk 6]
```diff
@@ -239,6 +245,43 @@ export const DefaultLayout = ({
...
+      return instances;
+    });
+    setCurrentNodeMetadata(currentNode);
+  }, [metadata.data, selectedItems]);
+
+  useEffect(() => {
+    if (!nodeIds.data) {
+      return;
+    }
+    setMetadataInstances((instances) => {
```
Comment (ainvoner) [inline apps/wing-console/console/ui/src/layout/default-layout.tsx:275]: I would add a comment describing this logic.
If I understand correctly, this cleans up all removed resources from the `metadataInstances` array, right?

File: apps/wing-console/console/ui/src/layout/default-layout.tsx [hunk 7]
```diff
@@ -367,27 +410,42 @@ export const DefaultLayout = ({
...
-                          node={metadata.data.node}
-                          inbound={metadata.data.inbound}
-                          outbound={metadata.data.outbound}
-                          onConnectionNodeClick={(path) => {
-                            expand(path);
-                            setSelectedItems([path]);
-                          }}
-                        />
-                      )}
+                      {Object.values(metadataInstances).map((instance) => (
```
Comment (ainvoner) [inline apps/wing-console/console/ui/src/layout/default-layout.tsx:423]: why do we need to load into the DOM everything and not only the one that is currently selected?

Comment (polamoros) [reply apps/wing-console/console/ui/src/layout/default-layout.tsx:425]: That's the trick to preserve the input's state for each opened node. Loading only the selected node is what causes the inputs to be 'reset' every time you switch from one node to another.

Comment (ainvoner) [reply apps/wing-console/console/ui/src/layout/default-layout.tsx:425]: but this is why we now have the instanceMetadata.data, no?
when we select a node, set also set the metadata data of this node (populate the relevant metadata fields)

Comment (polamoros) [reply apps/wing-console/console/ui/src/layout/default-layout.tsx:425]: One thing is the metadata (node info) and the other the inputs the user fills in (for the api the request type, the endpoint...) 🤔.

Comment (ainvoner) [reply apps/wing-console/console/ui/src/layout/default-layout.tsx:425]: wouldn't it be a better solution to have a ResourceMetadataContext in our app that will have the user input for each resource? this way, once a resource is selected, its' user input can be fetched and displayed.
I think it will be better than to have all resources metadata panels and their content on the DOM

Comment (skyrpex) [reply apps/wing-console/console/ui/src/layout/default-layout.tsx:425]: If the component is removed from the DOM, we'd lose their state (which is what we really want to keep).

An alternative to this is to build a more complex set of hooks that preserve the state for us.

File: apps/wing-console/console/design-system/src/key-value-list.tsx [hunk 8]
```diff
@@ -11,8 +11,14 @@ export interface KeyValueItem {
   value: string;
 }
 
-export const useKeyValueList = () => {
-  const [items, setItems] = useState<KeyValueItem[]>([]);
+export interface UseKeyValueListOptions {
+  useExternalState?: typeof useState;
```
Comment (skyrpex) [inline apps/wing-console/console/design-system/src/key-value-list.tsx:15]: I think the name `useState` makes more sense. `useExternalState` sounds like it's asking for a boolean, while `useState` is a well known method. Also, since `useExternalState` is a prop that you pass from the outside, the "external" part of the name is redundant.

File: apps/wing-console/console/ui/src/services/use-explorer.tsx [hunk 9]
```diff
@@ -56,15 +57,17 @@ export const useExplorer = () => {
...
-    setSelectedNode.mutate({
-      resourcePath: "root",
-    });
-  }, [tree.data, setItems]);
+    if (!selectedNode.data || !nodeIds.data?.includes(selectedNode.data)) {
+      setSelectedNode.mutate({
+        resourcePath: "root",
+      });
+    }
+  }, [tree.data, setItems, nodeIds.data]);
```
Comment (skyrpex) [inline apps/wing-console/console/ui/src/services/use-explorer.tsx:70]: I feel like this should become two different effects. Otherwise, `setItems` and `createTreeMenuItemFromExplorerTreeItem` will be called more than once if I'm not mistaken.

File: apps/wing-console/packages/use-persistent-state/src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  useContext,
+  useEffect,
+  useRef,
+  MutableRefObject,
+  useState,
+} from "react";
+
+const PersistentStateContext = createContext<{
+  state: MutableRefObject<Map<string, any[]>>;
+}>(undefined!);
```
Comment (skyrpex) [inline apps/wing-console/packages/use-persistent-state/src/index.tsx:15]: If we do `undefined!`, at some point we may end up trying to use `createPersistentState` without using the `PersistentStateProvider`, and it will fail silently.

It's better to:

```suggestion
const PersistentStateContext = createContext<{
  state: MutableRefObject<Map<string, any[]>>;
} | undefined>(undefined);
```

And create an auxiliary `usePersistentState` method that throws if the context is undefined.

**Review by skyrpex — CHANGES_REQUESTED** [review]

**Review by skyrpex — COMMENTED** [review]
Can you solve the new ESLint warnings?

File: apps/wing-console/packages/use-persistent-state/src/index.tsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+export const PersistentStateProvider = (props: PropsWithChildren) => {
+  const state = useRef(new Map<string, any[]>());
+  return (
+    <PersistentStateContext.Provider value={{ state }}>
+      {props.children}
+    </PersistentStateContext.Provider>
+  );
+};
+
+export const createPersistentState = (prefix: string) => {
```
Comment (ainvoner) [inline apps/wing-console/packages/use-persistent-state/src/index.tsx:39]: the prefix is actually the unique id for the state, right? why call it a prefix?

**Comment by polamoros** [comment]
> ounter state after hot reload?

My idea was to only save user inputs and not data that is loaded from the server.

Do you think we should save the other info as well?

**Comment by polamoros** [comment]
> @polamoros is it intentional not to save a counter state after hot reload? same question for bucket files that were already uploaded

My idea was to only save user inputs and not data that is loaded from the server.

Do you think we should save the other info as well?

File: apps/wing-console/packages/use-persistent-state/src/index.tsx (hunk 11 above)
Comment (polamoros) [reply apps/wing-console/packages/use-persistent-state/src/index.tsx]: true! it make no sense now.

fixed

Comment (skyrpex) [reply apps/wing-console/packages/use-persistent-state/src/index.tsx]: Initially it make sense but yeah, we should name it key or id.

**Comment by ainvoner** [comment]
regarding the counter - I agree there is no need to save the state.
regarding the bucket - i'm not sure. Users may want to have their already populated date in the bucket throughout hot reloading... on the other hand, if I had an issue with my date population to the bucket and now I fixed it, I will probably want to have a clean state.

**Comment by skyrpex** [comment]
> > @polamoros is it intentional not to save a counter state after hot reload? same question for bucket files that were already uploaded
> 
> My idea was to only save user inputs and not data that is loaded from the server.
> 
> Do you think we should save the other info as well?

I feel like we could use the `props.useState` everywhere in the components that have that prop. They shouldn't decide whether to use it in some instances or not.

**Comment by polamoros** [comment]
> regarding the counter - I agree there is no need to save the state. regarding the bucket - i'm not sure. Users may want to have their already populated date in the bucket throughout hot reloading... on the other hand, if I had an issue with my data population to the bucket and now I fixed it, I will probably want to have a clean state.

I wouldn't manage the resources state here.

We should distinguish between preserving the resource state and the view's state (user inputs).
If we want to preserve the resource state, it shouldn't be managed on the frontend. Preserving the state of resources when they are destroyed and recreated is more complex than simply storing some user inputs, I bet.

**Review by ainvoner — APPROVED** [review]

**Review by skyrpex — APPROVED** [review]
